inmoose/limma/decidetests.py

```diff
@@ -0,0 +1,269 @@
+# GNU General Public License for more details.
+
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <https://www.gnu.org/licenses/>.
+# -----------------------------------------------------------------------------
+
+# This file is based on the file 'R/decidetests.R' of the Bioconductor limma package (version 3.55.1).
+
+import numpy as np
+import pandas as pd
+import scipy
+
+from ..utils import cov2cor
+
+
+class TestResults(pd.DataFrame):
+    """
+    A matrix-based class for storing the results of simultaneous tests.
+
+    Instances of this class are usually created by :func:`decideTests`.
+
+    A :class:`TestResults` object is essentially a matrix with elements equal
+    to :math:`0`, :math:`1` or :math:`-1`. Zero represents acceptance of the
+    null hypothesis, :math:`1` indicates rejection in favor of the right tail
+    alternative and :math:`-1` indicates rejection in favor of the left tail
+    alternative.
+
+    :class:`TestResults` objects can be created by :code:`TestResults(results)`
+    where :code:`results` is a matrix.
+    """
+
+    _metadata = [
+        "levels",
+        "labels",
+    ]
+
+    @property
+    def _constructor(self):
+        return TestResults
+
+    @property
+    def _constructor_sliced(self):
+        return pd.Series
+
+    def __init__(self, df, *args, **kwargs):
+        super().__init__(df, *args, **kwargs)
+        self.levels = None
+        self.labels = None
+
+    def summary(self):
+        """
+        Summary of the test results
+
+        Returns
+        -------
+        pd.DataFrame
+            a summary table
+        """
+        levels = self.levels
+        if levels is None:
+            levels = [-1, 0, 1]
+        nlevels = len(levels)
+        tab = pd.DataFrame(np.zeros((nlevels, self.shape[1]), dtype=int))
+        labels = self.labels
+        if labels is None:
+            labels = levels
+        tab.index = labels
+        tab.columns = self.columns
+        for i in range(nlevels):
+            tab.iloc[i, :] = np.nan_sum(self == levels[i], axis=1)
+        return tab
+
+
+class FStat(np.ndarray):
+    """
+    Helper class to store F-statistics in a numpy array along with two degrees of freedom attributes
+
+    Subclassing ndarray is necessarily to add attributes to it.
+
+    Attributes
+    ----------
+    df1 : int
+        first degrees-of-freedom parameter of the F-statistics
+    df2 : int
+        second degrees-of-freedom parameter of the F-statistics
+    """
+
+    def __new__(cls, input_array, df1=None, df2=None):
+        obj = np.asarray(input_array).view(cls)
+        obj.df1 = df1
+        obj.df2 = df2
+        return obj
+
+    def __array_finalize__(self, obj):
+        if obj is None:
+            return
+        self.df1 = getattr(obj, "df1", None)
+        self.df2 = getattr(obj, "df2", None)
+
+
+def classifyTestsF(self, cor_matrix=None, df=np.inf, p_value=0.01, fstat_only=False):
```

Comment:
Is it ok to call a function argument self?

Reply:
`self` is just a name, it has no special meaning.
When calling a method on an object `obj.method(x,y,z)`, under the hood that call is transformed to `method(obj, x, y, z)`, hence the convention is to call `self` the first argument of methods.

Now you can also define methods outside of the class body, _e.g._ to break down a big class into smaller chunks, following this pattern:
```python
def f(self, x, y):
    return x + y + self.z

class MyClass:
    import f
```
In that case, calling `self` the first argument of `f` is a way to remind the reader that `f` is intended to be a method, even though it is defined out-of-scope.